### ISSUE TITLE
`typescript/no-floating-promises` behavior difference

### ISSUE THREAD

**Issue by connorshea** [oxc-project/tsgolint#958]
I had Claude create a reproduction of the problem here, same config for ESLint and Oxlint, but ESLint passes while Oxlint raises a violation: https://github.com/connorshea/oxlint-sandbox/tree/claude/update-linting-typescript-U4ayT

You can see 2 errors in Oxlint, only 1 in ESLint:

```
oxlint-sandbox % pnpm oxlint

  × [typescript(no-floating-promises)](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-floating-promises.html): Promises must be awaited, add void operator to ignore.
   ╭─[repro/positive.ts:3:1]
 2 │
 3 │ getPromise();
   · ─────────────
   ╰────
  help: The promise must end with a call to .catch, or end with a call to .then with a rejection handler, or be explicitly marked as ignored with the `void` operator.

  × [typescript(no-floating-promises)](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-floating-promises.html): Promises must be awaited, add void operator to ignore.
   ╭─[repro/case.ts:9:1]
 8 │
 9 │ getTest();
   · ──────────
   ╰────
  help: The promise must end with a call to .catch, or end with a call to .then with a rejection handler, or be explicitly marked as ignored with the `void` operator.

Found 0 warnings and 2 errors.
Finished in 107ms on 4 files with 84 rules using 18 threads.

oxlint-sandbox % pnpm eslint

/Users/connorshea/code/oxlint-sandbox/repro/positive.ts
  3:1  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises

✖ 1 problem (1 error, 0 warnings)
```

Code in question (https://github.com/connorshea/oxlint-sandbox/blob/claude/update-linting-typescript-U4ayT/repro/case.ts):
```ts
interface MyPromise<T> extends Promise<T> {
  __brand: string;
}

type Test = MyPromise<number> & { foo: 'bar' };

declare function getTest(): Test;

getTest();
```

This is the config that triggers the bug, I think something is wrong with `allowForKnownSafePromises`:
```json
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "options": {
    "typeAware": true
  },
  "plugins": ["typescript"],
  "rules": {
    "typescript/no-floating-promises": [
      "error",
      {
        "allowForKnownSafePromises": [
          { "from": "file", "name": "MyPromise" }
        ]
      }
    ]
  }
}
```
